README.md
# Graasp App: actions-viewer


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.2](https://github.com/graasp/graasp-app-actions-viewer/compare/v0.1.0...v0.1.2) (2020-07-07)

## 0.1.0 (2020-07-07)

### Features

- add chart with sample daily number of actions ([2218309](https://github.com/graasp/graasp-app-actions-viewer/commit/2218309dd30b6adee6e12f692f24d7946bf7251d)), closes [#1](https://github.com/graasp/graasp-app-actions-viewer/issues/1)

# Change Log


src/index.js
import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import Root from './components/Root';
import configureStore from './store/configureStore';
import 'react-dates/initialize';
import './index.css';
import 'react-dates/lib/css/_datepicker.css';

const root = document.getElementById('root');

const renderApp = (RootComponent, store) => {
  render(
    <Provider store={store}>
      <RootComponent />
    </Provider>,
    root,
  );
};

// render app to the dom
const { store, history } = configureStore();

renderApp(Root, store, history);

if (module.hot) {
  module.hot.accept('./components/Root', () => {
    // eslint-disable-next-line global-require
    const NextRoot = require('./components/Root').default;
    renderApp(NextRoot, store, history);
  });
}


src/config/propTypes.js
import PropTypes from 'prop-types';

const primaryShape = PropTypes.shape({
  main: PropTypes.string.isRequired,
  light: PropTypes.string.isRequired,
  dark: PropTypes.string.isRequired,
  contrastText: PropTypes.string.isRequired,
});

const overridesShape = PropTypes.shape({});

const greyShape = PropTypes.shape({
  50: PropTypes.string.isRequired,
  100: PropTypes.string.isRequired,
  200: PropTypes.string.isRequired,
  300: PropTypes.string.isRequired,
  400: PropTypes.string.isRequired,
  500: PropTypes.string.isRequired,
  600: PropTypes.string.isRequired,
  700: PropTypes.string.isRequired,
  800: PropTypes.string.isRequired,
  900: PropTypes.string.isRequired,
  A100: PropTypes.string.isRequired,
  A200: PropTypes.string.isRequired,
  A400: PropTypes.string.isRequired,
  A700: PropTypes.string.isRequired,
});

const h1Shape = PropTypes.shape({
  fontFamily: PropTypes.string.isRequired,
  fontWeight: PropTypes.number.isRequired,
  fontSize: PropTypes.string.isRequired,
  lineHeight: PropTypes.number.isRequired,
  letterSpacing: PropTypes.string.isRequired,
});

const buttonShape = PropTypes.shape({
  fontFamily: PropTypes.string.isRequired,
  fontWeight: PropTypes.number.isRequired,
  fontSize: PropTypes.string.isRequired,
  lineHeight: PropTypes.number.isRequired,
  letterSpacing: PropTypes.string.isRequired,
  textTransform: PropTypes.string.isRequired,
});

const themeType = PropTypes.shape({
  breakpoints: PropTypes.shape({
    keys: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    values: PropTypes.shape({
      xs: PropTypes.number.isRequired,
      sm: PropTypes.number.isRequired,
      md: PropTypes.number.isRequired,
      lg: PropTypes.number.isRequired,
      xl: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
  direction: PropTypes.string.isRequired,
  mixins: PropTypes.shape({
    toolbar: PropTypes.shape({
      minHeight: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
  overrides: overridesShape.isRequired,
  palette: PropTypes.shape({
    common: PropTypes.shape({
      black: PropTypes.string.isRequired,
      white: PropTypes.string.isRequired,
    }).isRequired,
    type: PropTypes.string.isRequired,
    primary: primaryShape.isRequired,
    secondary: PropTypes.shape({
      50: PropTypes.string.isRequired,
      100: PropTypes.string.isRequired,
      200: PropTypes.string.isRequired,
      300: PropTypes.string.isRequired,
      400: PropTypes.string.isRequired,
      500: PropTypes.string.isRequired,
      600: PropTypes.string.isRequired,
      700: PropTypes.string.isRequired,
      800: PropTypes.string.isRequired,
      900: PropTypes.string.isRequired,
      A100: PropTypes.string.isRequired,
      A200: PropTypes.string.isRequired,
      A400: PropTypes.string.isRequired,
      A700: PropTypes.string.isRequired,
      main: PropTypes.string.isRequired,
      light: PropTypes.string.isRequired,
      dark: PropTypes.string.isRequired,
      contrastText: PropTypes.string.isRequired,
    }).isRequired,
    error: primaryShape.isRequired,
    warning: primaryShape.isRequired,
    info: primaryShape.isRequired,
    success: primaryShape.isRequired,
    grey: greyShape.isRequired,
    contrastThreshold: PropTypes.number.isRequired,
    tonalOffset: PropTypes.number.isRequired,
    text: PropTypes.shape({
      primary: PropTypes.string.isRequired,
      secondary: PropTypes.string.isRequired,
      disabled: PropTypes.string.isRequired,
      hint: PropTypes.string.isRequired,
    }).isRequired,
    divider: PropTypes.string.isRequired,
    background: PropTypes.shape({
      paper: PropTypes.string.isRequired,
      default: PropTypes.string.isRequired,
    }).isRequired,
    action: PropTypes.shape({
      active: PropTypes.string.isRequired,
      hover: PropTypes.string.isRequired,
      hoverOpacity: PropTypes.number.isRequired,
      selected: PropTypes.string.isRequired,
      selectedOpacity: PropTypes.number.isRequired,
      disabled: PropTypes.string.isRequired,
      disabledBackground: PropTypes.string.isRequired,
      disabledOpacity: PropTypes.number.isRequired,
      focus: PropTypes.string.isRequired,
      focusOpacity: PropTypes.number.isRequired,
      activatedOpacity: PropTypes.number.isRequired,
    }).isRequired,
    default: greyShape.isRequired,
  }).isRequired,
  props: overridesShape.isRequired,
  shadows: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  typography: PropTypes.shape({
    htmlFontSize: PropTypes.number.isRequired,
    fontFamily: PropTypes.string.isRequired,
    fontSize: PropTypes.number.isRequired,
    fontWeightLight: PropTypes.number.isRequired,
    fontWeightRegular: PropTypes.number.isRequired,
    fontWeightMedium: PropTypes.number.isRequired,
    fontWeightBold: PropTypes.number.isRequired,
    h1: h1Shape.isRequired,
    h2: h1Shape.isRequired,
    h3: h1Shape.isRequired,
    h4: h1Shape.isRequired,
    h5: h1Shape.isRequired,
    h6: h1Shape.isRequired,
    subtitle1: h1Shape.isRequired,
    subtitle2: h1Shape.isRequired,
    body1: h1Shape.isRequired,
    body2: h1Shape.isRequired,
    button: buttonShape.isRequired,
    caption: h1Shape.isRequired,
    overline: buttonShape.isRequired,
    useNextconstiants: PropTypes.bool,
  }).isRequired,
  shape: PropTypes.shape({
    borderRadius: PropTypes.number.isRequired,
  }).isRequired,
  transitions: PropTypes.shape({
    easing: PropTypes.shape({
      easeInOut: PropTypes.string.isRequired,
      easeOut: PropTypes.string.isRequired,
      easeIn: PropTypes.string.isRequired,
      sharp: PropTypes.string.isRequired,
    }).isRequired,
    duration: PropTypes.shape({
      shortest: PropTypes.number.isRequired,
      shorter: PropTypes.number.isRequired,
      short: PropTypes.number.isRequired,
      standard: PropTypes.number.isRequired,
      complex: PropTypes.number.isRequired,
      enteringScreen: PropTypes.number.isRequired,
      leavingScreen: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
  zIndex: PropTypes.shape({
    mobileStepper: PropTypes.number.isRequired,
    speedDial: PropTypes.number.isRequired,
    appBar: PropTypes.number.isRequired,
    drawer: PropTypes.number.isRequired,
    modal: PropTypes.number.isRequired,
    snackbar: PropTypes.number.isRequired,
    tooltip: PropTypes.number.isRequired,
  }).isRequired,
  status: PropTypes.shape({
    danger: PropTypes.shape({
      background: greyShape.isRequired,
      color: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
}).isRequired;

// eslint-disable-next-line import/prefer-default-export
export const MATERIAL_UI_THEME_TYPE = themeType;


src/components/modes/teacher/TeacherView.js
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withStyles, withTheme } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Fab from '@material-ui/core/Fab';
import SettingsIcon from '@material-ui/icons/Settings';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import ReactMomentProptypes from 'react-moment-proptypes';
import { withTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import moment from 'moment';
import { extendMoment } from 'moment-range';
import { DateRangePicker } from 'react-dates';
import { Line } from '@nivo/line';
import './TeacherView.css';
import {
  deleteAppInstanceResource,
  openSettings,
  patchAppInstanceResource,
  postAppInstanceResource,
} from '../../../actions';
import { getUsers } from '../../../actions/users';
import { addQueryParamsToUrl } from '../../../utils/url';
import Settings from './Settings';
import { MATERIAL_UI_THEME_TYPE } from '../../../config/propTypes';
import { NIVO_LINE_MARGIN } from '../../../config/styles';
import { DATE_FROMAT } from '../../../config/formats';

export class TeacherView extends Component {
  static propTypes = {
    t: PropTypes.func.isRequired,
    dispatchOpenSettings: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      root: PropTypes.string,
      table: PropTypes.string,
      main: PropTypes.string,
      button: PropTypes.string,
      message: PropTypes.string,
      fab: PropTypes.string,
      chartTitle: PropTypes.string,
    }).isRequired,
    dispatchGetUsers: PropTypes.func.isRequired,
    // eslint-disable-next-line react/require-default-props
    theme: MATERIAL_UI_THEME_TYPE,
    actions: PropTypes.arrayOf(
      PropTypes.shape({
        date: ReactMomentProptypes.momentObj,
        count: PropTypes.number,
      }),
    ),
    minDate: ReactMomentProptypes.momentObj,
    maxDate: ReactMomentProptypes.momentObj,
  };

  static defaultProps = {
    actions: [],
    minDate: moment.utc(),
    maxDate: moment.utc(),
  };

  static styles = theme => ({
    root: {
      width: '100%',
      marginTop: theme.spacing(3),
      overflowX: 'auto',
    },
    main: {
      textAlign: 'center',
      margin: theme.spacing(),
    },
    button: {
      marginTop: theme.spacing(3),
    },
    table: {
      minWidth: 700,
    },
    message: {
      padding: theme.spacing(),
      backgroundColor: theme.status.danger.background[500],
      color: theme.status.danger.color,
      marginBottom: theme.spacing(2),
    },
    fab: {
      margin: theme.spacing(),
      position: 'fixed',
      bottom: theme.spacing(2),
      right: theme.spacing(2),
    },
    chartTitle: {
      marginBottom: theme.spacing(2),
    },
  });

  state = {
    fromDate: moment.utc(),
    toDate: moment.utc(),
  };

  constructor(props) {
    super(props);
    const { dispatchGetUsers } = this.props;
    dispatchGetUsers();
  }

  componentDidUpdate({ minDate: oldMinDate, maxDate: oldMaxDate }) {
    const { minDate, maxDate } = this.props;
    if (minDate !== oldMinDate) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        fromDate: minDate,
      });
    }
    if (maxDate !== oldMaxDate) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        toDate: maxDate,
      });
    }
  }

  handleActionsDateRange = () => {
    const { fromDate, toDate } = this.state;
    const { actions } = this.props;
    const rangedMoment = extendMoment(moment);
    const start = rangedMoment.utc(fromDate);
    const end = rangedMoment.utc(toDate);
    const range = rangedMoment.range(start, end);
    return Array.from(range.by('days')).map(day => ({
      x: day.format(DATE_FROMAT),
      y: actions.find(({ date }) => date.isSame(day, 'days'))?.count ?? 0,
    }));
  };

  render() {
    // extract properties from the props object
    const {
      // this property allows us to do styling and is injected by withStyles
      classes,
      // this property allows us to do translations and is injected by i18next
      t,
      dispatchOpenSettings,
      theme,
    } = this.props;
    const { fromDate, toDate, focusedInput } = this.state;
    const actionsDateRange = this.handleActionsDateRange();
    return (
      <>
        <Grid container spacing={0}>
          <Grid item xs={12} className={classes.main}>
            <Paper className={classes.message}>
              {t(
                'This is the teacher view. Switch to the student view by clicking on the URL below.',
              )}
              <a href={addQueryParamsToUrl({ mode: 'student' })}>
                <pre>
                  {`${window.location.host}/${addQueryParamsToUrl({
                    mode: 'student',
                  })}`}
                </pre>
              </a>
            </Paper>
            <Paper className={classes.root}>
              <Typography
                variant="h5"
                color="inherit"
                className={classes.chartTitle}
              >
                {t('Number of Daily Actions')}
              </Typography>
              <DateRangePicker
                startDate={fromDate}
                startDateId="1"
                endDate={toDate}
                endDateId="2"
                onDatesChange={({ startDate, endDate }) => {
                  this.handleActionsDateRange();
                  this.setState({ fromDate: startDate, toDate: endDate });
                }}
                focusedInput={focusedInput}
                onFocusChange={input => this.setState({ focusedInput: input })}
                isOutsideRange={() => false}
                small
              />
              <Line
                data={[
                  {
                    id: 1,
                    data: actionsDateRange,
                  },
                ]}
                xScale={{
                  type: 'time',
                  format: '%d/%m/%Y',
                  useUTC: true,
                  precision: 'day',
                }}
                yScale={{
                  type: 'linear',
                  stacked: false,
                }}
                axisLeft={{
                  orient: 'left',
                  legendPosition: 'middle',
                  legend: t('Number of Actions'),
                  legendOffset: -35,
                }}
                curve="linear"
                axisBottom={{
                  orient: 'bottom',
                  format: '%d/%m/%y',
                  legend: 'Time',
                  legendOffset: 40,
                  tickRotation: -25,
                  legendPosition: 'middle',
                }}
                colors={theme.palette.primary.main}
                enablePoints={false}
                xFormat="time:%d/%m/%Y"
                enableSlices="x"
                sliceTooltip={({ slice }) => {
                  return (
                    <div
                      style={{
                        background: 'white',
                        padding: '9px 12px',
                        border: '1px solid #ccc',
                      }}
                    >
                      {slice.points.map(
                        ({
                          data: { xFormatted, yFormatted },
                          serieColor,
                          id,
                        }) => (
                          <>
                            <div>
                              {`${t('Date')}`}
                              :&nbsp;
                              {xFormatted}
                            </div>
                            <div>
                              {`${t('Number of Actions')}`}
                              :&nbsp;
                              <span
                                key={id}
                                style={{
                                  color: serieColor,
                                }}
                              >
                                {yFormatted}
                              </span>
                            </div>
                          </>
                        ),
                      )}
                    </div>
                  );
                }}
                margin={NIVO_LINE_MARGIN}
                width={900}
                height={400}
              />
            </Paper>
          </Grid>
        </Grid>
        <Settings />
        <Fab
          color="primary"
          aria-label={t('Settings')}
          className={classes.fab}
          onClick={dispatchOpenSettings}
        >
          <SettingsIcon />
        </Fab>
      </>
    );
  }
}

// get the app instance resources that are saved in the redux store
const mapStateToProps = ({ appInstanceResources, action }) => {
  const actionsPerDay = action.content.reduce((acc, obj) => {
    const spaceDate = moment(obj.createdAt);
    const index = acc.findIndex(({ date }) => spaceDate.isSame(date, 'day'));
    if (index > 0) {
      acc[index] = {
        date: spaceDate,
        count: acc[index].count + 1,
      };
    } else {
      acc.push({
        date: spaceDate,
        count: 1,
      });
    }
    return acc;
  }, []);
  const sortedActions = actionsPerDay.sort(({ date: dateA }, { date: dateB }) =>
    moment(dateA).isBefore(dateB),
  );
  const {
    0: minDateAction,
    [actionsPerDay.length - 1]: maxDateAction,
  } = actionsPerDay;
  return {
    appInstanceResources: appInstanceResources.content,
    actions: sortedActions,
    minDate: minDateAction?.date,
    maxDate: maxDateAction?.date,
  };
};

// allow this component to dispatch a post
// request to create an app instance resource
const mapDispatchToProps = {
  dispatchGetUsers: getUsers,
  dispatchPostAppInstanceResource: postAppInstanceResource,
  dispatchPatchAppInstanceResource: patchAppInstanceResource,
  dispatchDeleteAppInstanceResource: deleteAppInstanceResource,
  dispatchOpenSettings: openSettings,
};

const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TeacherView);

const StyledComponent = withStyles(TeacherView.styles)(ConnectedComponent);
const ThemedComponent = withTheme(StyledComponent);

export default withTranslation()(ThemedComponent);
